nilearn/image/image.py

```diff
@@ -842,7 +862,45 @@ def threshold_img(img, threshold, mask_img=None, copy=True):
                                            percentile_func=scoreatpercentile,
                                            name='threshold')
 
-    img_data[np.abs(img_data) < cutoff_threshold] = 0.
+    # Apply threshold
+    if two_sided:
+        img_data[np.abs(img_data) < cutoff_threshold] = 0.
+    else:
+        img_data[img_data < cutoff_threshold] = 0.
+
+    # Expand to 4D to support both 3D and 4D
+    expand_to_4d = img_data.ndim == 3
+    if expand_to_4d:
+        img_data = img_data[:, :, :, None]
+
+    # Perform cluster thresholding, if requested
+    if cluster_threshold > 0:
+        # Define array for 6-connectivity, aka NN1 or "faces"
+        conn_mat = np.zeros((3, 3, 3), int)
+        conn_mat[1, 1, :] = 1
+        conn_mat[1, :, 1] = 1
+        conn_mat[:, 1, 1] = 1
+
+        for i_vol in range(img_data.shape[3]):
+            vol_data = img_data[..., i_vol]
```

Comment:
Do you think we could put the code of the inner loop together with the connectivity matrix definition in a helper function?

It seems to me that this could then be parallelized but I might be missing something (or maybe it's fast enough we don't need to bother about this...). :thinking:

Reply:
I moved the inner loop into a new private function, but I didn't try parallelizing or anything.